### PR TITLE
VelocitySupport

### PR DESCRIPTION
**Pull request by fede1132**
I made AdvancedBan compatible with Velocity Powered.

List of things must be revised:
* ChatListenerVelocity.java
* ConnectionListenerVelocity.java

Known bugs:
* UUID Fetcher not working
* ~~Comments not saved into config file~~ Fixed
* ~~MySQL not working~~ (Working but plugin size is 7mb)

Things missing (due to Velocity "limitations")
* Plugin messaging

Thanks to:
@JavaFactoryDev for the storage library I used.

[Download beta version](https://drive.google.com/file/d/17D1J4IjJ-bXS9Pfyo1iw6pkv_PPgH-qb/view?usp=sharing)

### CODE REVIEW

**Comment by DevLeoko** [comment]
Looks good all in all, but:
- The plugin version should not be a hardcoded constant in the code. If velocity does not support reading the version from the plugin.yml or bungee.yml then please try something like this: https://stackoverflow.com/questions/3697449/retrieve-version-from-maven-pom-xml-in-code/41791885
- Please try to do something about the plugin size, this would make AdvancedBan about 3x it's original size :(
- Why did you add the `MySQL.yml`? All MySQL Data are available in the config.yml
- Do you know why the UUID Fetcher is not working? Can you do something about that? Seems to me like a major issue

**Comment by fede1132** [comment]
> Looks good all in all, but:
> 
> * The plugin version should not be a hardcoded constant in the code. If velocity does not support reading the version from the plugin.yml or bungee.yml then please try something like this: https://stackoverflow.com/questions/3697449/retrieve-version-from-maven-pom-xml-in-code/41791885
> * Please try to do something about the plugin size, this would make AdvancedBan about 3x it's original size :(
> * Why did you add the `MySQL.yml`? All MySQL Data are available in the config.yml
> * Do you know why the UUID Fetcher is not working? Can you do something about that? Seems to me like a major issue

1. Ok, I'm gonna do this
1. Forgot during some tests
1. I'm trying to understand this..

**Comment by DevLeoko** [comment]
I don't really see why you use that particular storage library. If it's too big and Velocity don't have their own YAML parser then just use a different one like SnakeYAML which is just 300kb

**Comment by JavaFactoryDev** [comment]
Yes and LightningStorage is only 291 kb. https://prnt.sc/u0m0wk
Just saw I got mentioned.

**Comment by fede1132** [comment]
> I don't really see why you use that particular storage library. If it's too big and Velocity don't have their own YAML parser then just use a different one like SnakeYAML which is just 300kb

The file size is 7mb because Velocity does not have a built-in mysql-connector so I need to shade it into the jar.
And anyway Leonhard's library is the best library I've ever seen, it automatically reloads config when these are updated.

**Comment by DevLeoko** [comment]
Oh sorry, my bad, the message above sounded a bit like the storage library was responsible for the large size.
In that case we should maybe not include the Velocity build in the bundled version but make it build separately.

Also, let me know when there are any updates on the UUID situation ^^

**Comment by fede1132** [comment]
I think the problem Is with outed API urls. I'm trying to add Support for playerdb that Is a cached player API for Steam Games, Minecraft and Valve games. Powered by Steam

**Comment by DevLeoko** [comment]
> In that case we should maybe not include the Velocity build in the bundled version but make it build separately.

Have you given that a thought?

**Comment by fede1132** [comment]
> > In that case we should maybe not include the Velocity build in the bundled version but make it build separately.
> 
> Have you given that a thought?

I'm busy in these days I'll di this maybe this weekend

File: core/src/main/resources/MySQL.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+MySQL:
+  IP: localhost
```
Comment (A248) [inline core/src/main/resources/MySQL.yml:2]: Why is the MySQL.yml brought back? I thought those settings were moved into config.yml

File: velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        } catch (Exception exc) {
+            return null;
+        }
+    }
+
+    @Override
+    public String getVersion() {
+        String version = "Unknown";
+        try (InputStream stream = getClass().getClassLoader().getResourceAsStream("velocity-plugin.json"); InputStreamReader read = new InputStreamReader(stream)) {
+            JsonObject obj = new JsonParser().parse(read).getAsJsonObject();
```
Comment (A248) [inline velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java:80]: There has to be a better way to do this. This method does IO and parses Json every time it is called. Why not use Velocity's methods on PluginContainer?

File: velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+        messages = LightningBuilder.fromFile(messageFile).addInputStreamFromResource(messageFile.getName()).setConfigSettings(ConfigSettings.PRESERVE_COMMENTS).createYaml();
+        layouts = LightningBuilder.fromFile(layoutFile).addInputStreamFromResource(layoutFile.getName()).setConfigSettings(ConfigSettings.PRESERVE_COMMENTS).createYaml();
+        if (mysqlFile.exists()) {
+            mysql = LightningBuilder.fromFile(mysqlFile).setConfigSettings(ConfigSettings.PRESERVE_COMMENTS).createYaml();
+        } else mysql = config;
+    }
+
+    @Override
+    public String getFromUrlJson(String url, String key) {
+        try {
```
Comment (A248) [inline velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java:57]: Although I know this is somewhat outside the scope of this PR, this method's body is much duplicated. Rather than write it a 3rd time, perhaps there is a better way to consolidate all of it?

File: velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMain.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+@Plugin(
+        id = VelocityMain.VelocityPluginInfo.PLUGIN_ID,
+        name = VelocityMain.VelocityPluginInfo.PLUGIN_NAME,
+        version = "${project.version}",
+        authors = {"Leoko", "Fede1132(Added velocity support)"}
+)
+public class VelocityMain {
+    public static class VelocityPluginInfo {
+        public static final String PLUGIN_ID = "advancedban";
```
Comment (A248) [inline velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMain.java:26]: Why are this class and its constants public?

File: velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    @Override
+    public String getInternUUID(String player) {
+        Optional<Player> optionalPlayer = getPlugin().server.getPlayer(player);
+        return optionalPlayer.map(value -> value.getUniqueId().toString().replace("-", "")).orElse(null);
+    }
+
+    @Override
+    public boolean hasPerms(Object player, String perms) {
+        if (player instanceof Player) {
+            return ((Player) player).hasPermission(perms);
```
Comment (A248) [inline velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java:178]: This is a completely unnecessary instanceof/cast. The way AdvancedBan is made, you need only `return ((CommandSource) player).hasPermission(perms)`

File: velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+    }
+
+    @Override
+    public boolean isOnline(String name) {
+        return getPlugin().server.getPlayer(name).isPresent();
+    }
+
+    @Override
+    public Object getPlayer(String name) {
+        return getPlugin().server.getPlayer(name).get();
```
Comment (A248) [inline velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java:196]: Using `Optional#get` changes the behaviour of this method from other MethodInterface implementations, making this one throw NoSuchElementException for nonexistent values. Also, the overridden signature can be changed to `Player`.

File: velocity/src/main/java/me/leoko/advancedban/velocity/listener/CommandReceiverVelocity.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+            args[0] = (VelocityMain.get().server.getPlayer(args[0]).isPresent() ? VelocityMain.get().server.getPlayer(args[0]).get().getUsername() : args[0]);
+        }
+        CommandManager.get().onCommand(commandSource, cmd, args);
+    }
+
+    @Override
+    public List<String> suggest(CommandSource source, @NonNull String[] args) {
+        final me.leoko.advancedban.utils.Command command = me.leoko.advancedban.utils.Command.getByName(cmd.substring(1));
+        if (command != null && source instanceof Player) {
+            if (command.getPermission() == null || Universal.get().getMethods().hasPerms(source, command.getPermission())) {
```
Comment (A248) [inline velocity/src/main/java/me/leoko/advancedban/velocity/listener/CommandReceiverVelocity.java:32]: These can be collapsed.

File: velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMain.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        public static final String PLUGIN_NAME = "AdvancedBan";
+    }
+    private static VelocityMain instance;
+    public final ProxyServer server;
+    public final Logger logger;
```
Comment (A248) [inline velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMain.java:31]: I would discourage using public fields. Make a couple getters instead.

File: velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+
+    @Override
+    public void setCommandExecutor(String cmd, TabCompleter tabCompleter) {
+        getPlugin().server.getCommandManager().register(new CommandReceiverVelocity(cmd), cmd);
+    }
+
+    @Override
+    public void sendMessage(Object player, String msg) {
+        if (player instanceof Player) {
+            ((Player) player).sendMessage(TextComponent.of(msg));
```
Comment (A248) [inline velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java:139]: Likewise with #hasPermission, this is an unnecessary cast. Also, are you sure `TextComponent.of` is the correct method? I thought the LegacyComponentSerializer was supposed to be used, although it's possible TextComponent.of will work for now.

File: velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+import com.google.gson.JsonElement;
+import com.google.gson.JsonNull;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.velocitypowered.api.command.CommandSource;
+import com.velocitypowered.api.proxy.Player;
+import de.leonhard.storage.LightningBuilder;
+import de.leonhard.storage.Yaml;
+import de.leonhard.storage.internal.settings.ConfigSettings;
+import jdk.nashorn.internal.parser.JSONParser;
```
Comment (A248) [inline velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java:12]: Why is this here?

File: core/src/main/resources/MySQL.yml (hunk 1 above)
Comment (fede1132) [reply core/src/main/resources/MySQL.yml:2]: As I already said before.. the library I used require an InputStream to inizialize the file, but of you look at the First lines for VelocityMethods mysql file Will be loader ONLY if the file exists

File: velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java (hunk 2 above)
Comment (fede1132) [reply velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMethods.java:80]: Oh ok, my idea was different at begininng so I forgot to update this

File: velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMain.java (hunk 4 above)
Comment (fede1132) [reply velocity/src/main/java/me/leoko/advancedban/velocity/VelocityMain.java:26]: As I said the idea at begininng was different.. so I forgot to update it

**Comment by DevLeoko** [comment]
Velocity Support has been added though #434 (and this PR looks inactive) so I'll close this PR. Thanks though for your effort.
